src/Pact/Eval.hs

```diff
@@ -818,22 +818,56 @@ applyPact _ _ t _ = evalError' t "applyPact: invalid defpact body, expected list
 -- | Resume a pact, either as specified or as found in database.
 -- Expects a 'PactStep' to be populated in the environment.
 resumePact :: Info -> Maybe PactExec -> Eval e (Term Name)
-resumePact i pe = do
+resumePact i crossChainContinuation = do
 
   ps@PactStep{..} <- view eePactStep >>= (`maybe` pure)
     (evalError i "resumePact: no step in environment")
 
-  context <- case pe of
-    Just p -> return p
-    Nothing -> do
-      contextM <- readRow i Pacts _psPactId >>= (`maybe` pure)
-        (evalError i $ "resumePact: no previous execution found for: " <> pretty _psPactId)
-
-      case contextM of
-        Nothing -> evalError i $ "resumePact: pact completed: " <> pretty _psPactId
-        Just c -> return c
-
-  resumePactExec i ps context
+  -- query for previous exec
+  dbState <- readRow i Pacts _psPactId
+
+  -- validate db state
+  let proceed = resumePactExec i ps
+  case (dbState,crossChainContinuation) of
+
+    -- Terminated pact in db: always fail
+    (Just Nothing,_) ->
+      evalError i $ "resumePact: pact completed: " <> pretty _psPactId
+
+    -- Nothing in db, Nothing cross-chain continuation: fail
+    (Nothing,Nothing) ->
+      evalError i $ "resumePact: no previous execution found for: " <> pretty _psPactId
+
+    -- Nothing in db, Just cross-chain continuation: proceed with cross-chain
+    (Nothing,Just ccExec) -> proceed ccExec
+
+    -- Active db record, Nothing cross-chain continuation: proceed with db
+    (Just (Just dbExec),Nothing) -> proceed dbExec
+
+    -- Active db record, cross-chain continuation:
+    -- TODO: this is a fishy case, consider disallowing.
+    -- TODO: if we do decide to allow, cover in tests.
+    -- For now, a valid possibility iff this is a flip-flop from another chain, e.g.
+    --   0. This chain: start pact
+    --   1. Other chain: continue pact
+    --   2. This chain: continue pact
+    (Just (Just dbExec),Just ccExec)
```

Comment:
This is a valid possibility if and only if the user has supplied an optional proof for same-chain steps. In which case, we need to verify everything in the pact execs on `==`. There could potentially be a use-case here, but it does seem redundant to want SPV proof for same-chain steps which are already guaranteed by the blockchain structure. I'm leaning towards "only in the case of cross-chain" on this one just to limit our number of edge cases.

Reply:
Per discussion, allowed if skipping at least one, and validated continuation and step count (the others are ignored in `resumePactExec`